core/transaction_pool/impl/transaction_pool_impl.cpp

```diff
@@ -0,0 +1,190 @@
+        ready_queue_.insert({rtx.hash, rtx});
+        auto old_it = it;
+        it++;
+        waiting_queue_.erase(old_it);
+      } else {
+        it++;
+      }
+    }
+  }
+
+  void TransactionPoolImpl::updateUnlockingTransactions(
+      const ReadyTransaction &rtx) {
+    for (auto &tag : rtx.requires) {
+      if (auto it = provided_tags_.find(tag); it != provided_tags_.end()) {
+        if (auto ready_it = ready_queue_.find(it->second);
+            ready_it != ready_queue_.end()) {
+          auto &unlocking_transaction = ready_it->second;
+          unlocking_transaction.unlocks.push_back(rtx.hash);
+        }
+      }
+    }
+  }
+
+  std::vector<primitives::Transaction> TransactionPoolImpl::enforceLimits() {
+    auto [ready_num, waiting_num] = getStatus();
+    int64_t dr = ready_num - limits_.max_ready_num;
+    int64_t dw = waiting_num - limits_.max_waiting_num;
+    std::vector<Transaction> removed;
+    removed.reserve((dr > 0 ? dr : 0) + (dw > 0 ? dw : 0));
+    while (ready_num > limits_.max_ready_num) {
+      removed.push_back(ready_queue_.begin()->second);
+      ready_queue_.erase(ready_queue_.begin());
+      ready_num--;
+    }
+    while (waiting_num > limits_.max_waiting_num) {
+      removed.push_back(waiting_queue_.front());
+      waiting_queue_.pop_front();
+      waiting_num--;
+    }
+    return removed;
+  }
+
+  TransactionPoolImpl::Status TransactionPoolImpl::getStatus() const {
+    return Status{ready_queue_.size(), waiting_queue_.size()};
+  }
+
+  std::vector<Transaction> TransactionPoolImpl::pruneTags(
+      const primitives::BlockId &at, const primitives::TransactionTag &tag,
+      const std::vector<common::Hash256> &known_imported_hashes) {
+    std::list<primitives::TransactionTag> to_remove{tag};
+    std::vector<Transaction> removed;
+
+    while (not to_remove.empty()) {
+      auto tag = to_remove.front();
+      to_remove.pop_front();
+      auto hash = provided_tags_.at(tag);
+      provided_tags_.erase(tag);
+      auto tx = ready_queue_.at(hash);
+      ready_queue_.erase(hash);
+
+      auto find_previous = [this, &tx](primitives::TransactionTag const &tag)
+          -> std::optional<std::vector<primitives::TransactionTag>> {
+        auto prev_hash = provided_tags_.at(tag);
+        auto &tx2 = ready_queue_[prev_hash];
+        tx2.unlocks.remove_if([&tx](auto &t) { return t == tx.hash; });
+        // We eagerly prune previous transactions as well.
+        // But it might not always be good.
+        // Possible edge case:
+        // - tx provides two tags
+        // - the second tag enables some subgraph we don't know of yet
+        // - we will prune the transaction
+        // - when we learn about the subgraph it will go to future
+        // - we will have to wait for re-propagation of that transaction
+        // Alternatively the caller may attempt to re-import these transactions.
+        if (tx2.unlocks.empty()) {
+          return std::optional(tx2.provides);
```

Comment:
std::make_optional

Reply:
What is the difference with working deduction guides?

Reply:
there is no difference, i guess. 
this one is just more natural.

like as if you would create shared_ptr by doing `return std::shared_ptr(ptr)`

Reply:
As for me it is natural to use make_shared when I allocate an object in memory, but here I just wrap an existing object in an optional, so if you don't mind I'll keep it the way it is.